### PR TITLE
Bigquery Native Tables support

### PR DESCRIPTION
**Pull request by AmandeepSinghCS**

### CODE REVIEW

File: clouddq/classes/dq_configs_cache.py
Comment (nauh) [inline clouddq/classes/dq_configs_cache.py:473]: Remove this sentence? This seems a bit unrelated if the error is about missing "project" or "datasets".

File: clouddq/classes/dq_configs_cache.py
Comment (nauh) [inline clouddq/classes/dq_configs_cache.py:476]: Just `entity_uri.configs_dict.get("projects")`, without adding another method in dq_entity_uri.py?

File: clouddq/integration/bigquery/bigquery_client.py
Comment (nauh) [inline clouddq/integration/bigquery/bigquery_client.py:251]:
```suggestion
                time_partitioning = table_ref.time_partitioning
                if time_partitioning:
                    partition_field = {}
                    time_partitioning_type = time_partitioning.type_
                    time_partitioning_field = time_partitioning.field
                    if not time_partitioning_field:
                        partition_fields = "_PARTITIONTIME"
                    else:
                        partition_field["name"] = time_partitioning_field
                        partition_field["type"] = columns[
                            time_partitioning_field.upper()
                        ]["type"]
                        partition_field["partitioning_type"] = time_partitioning_type
                        partition_fields.append(partition_field)

                range_partitioning = table_ref.range_partitioning
                if range_partitioning:
                    raise NotImplementedError(
                        logger.debug(f"Range Partitioning is not supported currently.")
                    )
```

File: clouddq/integration/bigquery/bigquery_client.py
Comment (nauh) [inline clouddq/integration/bigquery/bigquery_client.py:246]: Can there be multiple partition fields for a table? (wondering why it's a list)

File: clouddq/integration/bigquery/bigquery_client.py
Comment (nauh) [inline clouddq/integration/bigquery/bigquery_client.py:267]: Can this error be thrown from multiple lines in the try block or only from one line? If it's the latter, can we only put that line in a try-except block?

File: tests/integration/test_dataplex_metadata_uri_templates.py
Comment (nauh) [inline tests/integration/test_dataplex_metadata_uri_templates.py:171]:
```suggestion
            output = output.replace("CAST(NULL AS STRING) AS dataplex_asset_id,", ASSET_ID_REP)

            output = output.replace(rule_binding_id, "<rule_binding_id>")
            output = re.sub(RE_NEWLINES, '\n', output).strip()
            output = re.sub(RE_CONFIGS_HASHSUM, CONFIGS_HASHSUM_REP, output)
            output = re.sub(RE_ASSET_ID, ASSET_ID_REP, output)
```

**Review by nauh — COMMENTED** [review]
Thanks Amandeep! I left a few minor comments.

File: clouddq/integration/bigquery/bigquery_client.py
Comment (AmandeepSinghCS) [reply clouddq/integration/bigquery/bigquery_client.py:246]: Hive supports multiple partition fields, Metadata API returns a List for Partition Fields . For consistency I have kept the same as we transform them to entity object which can be GCS or Bigquery.

File: clouddq/classes/dq_configs_cache.py
Comment (AmandeepSinghCS) [reply clouddq/classes/dq_configs_cache.py:473]: done

File: clouddq/integration/bigquery/bigquery_client.py
Comment (AmandeepSinghCS) [reply clouddq/integration/bigquery/bigquery_client.py:251]: incorporated the suggestion.

File: clouddq/classes/dq_configs_cache.py
Comment (AmandeepSinghCS) [reply clouddq/classes/dq_configs_cache.py:476]: done

File: clouddq/integration/bigquery/bigquery_client.py
Comment (AmandeepSinghCS) [reply clouddq/integration/bigquery/bigquery_client.py:219]: it will be thrown from one line. incorporated the change.

File: clouddq/classes/dq_configs_cache.py
Comment (thinhha) [inline clouddq/classes/dq_configs_cache.py:475]: s/entity_uri.configs_dict.get("projects")/entity_uri.get_configs("projects")/

File: clouddq/classes/dq_entity.py
Comment (thinhha) [inline clouddq/classes/dq_entity.py:366]: this should take the bigquery entity_uri directly.

File: clouddq/classes/dq_configs_cache.py
Comment (thinhha) [inline clouddq/classes/dq_configs_cache.py:497]: please add entity_uri into error log

File: clouddq/integration/bigquery/bigquery_client.py
Comment (thinhha) [inline clouddq/integration/bigquery/bigquery_client.py:253]: please add more descriptive error message. e.g. BigQuery range partition for table <> is not supported.

**Review by thinhha — CHANGES_REQUESTED** [review]

File: clouddq/integration/bigquery/bigquery_client.py
Comment (AmandeepSinghCS) [reply clouddq/integration/bigquery/bigquery_client.py:253]: done

File: clouddq/classes/dq_configs_cache.py
Comment (AmandeepSinghCS) [reply clouddq/classes/dq_configs_cache.py:497]: done

File: clouddq/classes/dq_configs_cache.py
Comment (AmandeepSinghCS) [reply clouddq/classes/dq_configs_cache.py:483]: done

File: clouddq/classes/dq_configs_cache.py
Comment (AmandeepSinghCS) [reply clouddq/classes/dq_configs_cache.py:475]: done

File: clouddq/classes/dq_entity.py
Comment (AmandeepSinghCS) [reply clouddq/classes/dq_entity.py:366]: updated

File: clouddq/classes/dq_configs_cache.py
Comment (AmandeepSinghCS) [reply clouddq/classes/dq_configs_cache.py:490]: updated

File: clouddq/classes/dq_configs_cache.py
Comment (AmandeepSinghCS) [reply clouddq/classes/dq_configs_cache.py:446]: correct me if I am wrong or if I am missing something here but why do we need to call metadata API, if we are using BQ tables API to look for any table in the dataset that is assigned as a asset to dataplex or not.

File: clouddq/classes/dq_configs_cache.py
Comment (nauh) [inline clouddq/classes/dq_configs_cache.py:460]: "registered as" -> "attached to"

File: clouddq/classes/dq_configs_cache.py
Comment (nauh) [inline clouddq/classes/dq_configs_cache.py:555]: s/exists/exist

File: clouddq/classes/dq_configs_cache.py
Comment (AmandeepSinghCS) [reply clouddq/classes/dq_configs_cache.py:460]: updated this to=> Ensure the BigQuery dataset containing this table is attached as an asset in Dataplex

File: clouddq/classes/dq_configs_cache.py
Comment (AmandeepSinghCS) [reply clouddq/classes/dq_configs_cache.py:555]: updated

File: clouddq/classes/dq_configs_cache.py
Comment (thinhha) [inline clouddq/classes/dq_configs_cache.py:528]: this else clause is being run inside of the loop `for argument in required_arguments:`, which means that it is executed at least 3 times, and may fail if a uri has `projects` but not `dataset`.

Please take everything after the else clause out of this loop.

File: clouddq/classes/dq_entity_uri.py
Comment (thinhha) [inline clouddq/classes/dq_entity_uri.py:183]: maybe just `get_table_name` is sufficient? the entity_uri part if self-explanatory from the parent class.

File: clouddq/classes/dq_configs_cache.py
Comment (thinhha) [inline clouddq/classes/dq_configs_cache.py:494]: I think we have the same bug here as discussed earlier Amandeep. Can you please move this out of the for loop?

File: clouddq/classes/dq_entity_uri.py
Comment (thinhha) [reply clouddq/classes/dq_entity_uri.py]: have you changed this method name as well @AmandeepSinghCS ?

Comment (AmandeepSinghCS) [reply clouddq/classes/dq_entity_uri.py]: @thinhha yes this is renamed

File: clouddq/classes/dq_configs_cache.py
Comment (nauh) [inline clouddq/classes/dq_configs_cache.py:496]: Can we rename `client` to `dataplex_client`?

File: clouddq/classes/dq_configs_cache.py
Comment (nauh) [inline clouddq/classes/dq_configs_cache.py:535]: "exists in BigQuery"?

File: clouddq/classes/dq_configs_cache.py
Comment (nauh) [inline clouddq/classes/dq_configs_cache.py:528]: Is this check for the BigQuery native tables only, or BigQuery native tables and external tables? If the latter, then maybe just call it `bq_table_exists`?

On a related note, it's probably good to clarify what "Bigquery Native Tables Support" entails exactly (regarding external table support).

File: clouddq/classes/dq_configs_cache.py
Comment (nauh) [inline clouddq/classes/dq_configs_cache.py:544]: If `entity_id` can be retrieved from `entity_uri`, then we don't need to pass both `entity_id` and `entity_uri` to the `from_bq_entity_uri` method -- only `entity_uri` should suffice.

File: clouddq/integration/bigquery/bigquery_client.py
Comment (nauh) [inline clouddq/integration/bigquery/bigquery_client.py:252]: Should this be debug level logging or higher?

File: clouddq/integration/bigquery/bigquery_client.py
Comment (nauh) [inline clouddq/integration/bigquery/bigquery_client.py:240]: How is `partition_fields` used? I wonder how we distinguish it being a string (i.e. "_PARTITIONTIME") from it being an array.

File: clouddq/BUILD
Comment (nauh) [inline clouddq/BUILD:59]: Just curious, where do we actually use `markupsafe` lib from the code?

File: tests/resources/configs/rule_bindings/team-7-rule-bindings.yml
Comment (nauh) [inline tests/resources/configs/rule_bindings/team-7-rule-bindings.yml:17]: I wonder if we should use a public dataset or something created and owned by us.

- Theoretically, public dataset can change (? for example, adding a new column), which can then make our test fail. I don't think we should create such dependency for our tests.
- Public dataset does not provide full coverage of our test scenarios. For example, `partition_fields` is either `None` or `null` for all the testing cases in this PR. I think we should have at least a test case where it's not `None` or `null`.

File: clouddq/classes/dq_configs_cache.py
Comment (AmandeepSinghCS) [reply clouddq/classes/dq_configs_cache.py]: done

File: clouddq/classes/dq_configs_cache.py
Comment (AmandeepSinghCS) [reply clouddq/classes/dq_configs_cache.py:535]: done

File: clouddq/classes/dq_configs_cache.py
Comment (AmandeepSinghCS) [reply clouddq/classes/dq_configs_cache.py:528]: It can be native or external table created by dataplex. We can rename it to the bq_table_exists.

File: clouddq/classes/dq_configs_cache.py
Comment (AmandeepSinghCS) [reply clouddq/classes/dq_configs_cache.py:544]: updated the method definition

File: clouddq/integration/bigquery/bigquery_client.py
Comment (AmandeepSinghCS) [reply clouddq/integration/bigquery/bigquery_client.py:252]: This should be only message for NotImplementedError, logger is not needed. Thanks for catching this.

File: clouddq/BUILD
Comment (AmandeepSinghCS) [reply clouddq/BUILD:59]: we use jinja, jinja internally has a dependency on markupsafe.

File: clouddq/integration/bigquery/bigquery_client.py
Comment (AmandeepSinghCS) [reply clouddq/integration/bigquery/bigquery_client.py:240]: _PARTITIONTIME is  a pseudocolumn for ingestion-time partitioning. The value of this column is the ingestion time for each row, truncated to the partition boundary (such as hourly or daily). This will be the default partition field. In case of metadata for GCS entities, it may have multiple partition fields so a list is used. For custom hive partitioning.

File: tests/resources/configs/rule_bindings/team-7-rule-bindings.yml
Comment (AmandeepSinghCS) [reply tests/resources/configs/rule_bindings/team-7-rule-bindings.yml:17]: 1. Theoretically, public dataset can change (? for example, a new column added), which can then make our test fail. I don't think we should create such dependency for our tests.

> I have made a copy of public dataset in our project, so the table structure is not going to change until we make a change. So our tests will run independent of public dataset.

2.  Public dataset does not provide full coverage of our test scenarios. For example, partition_fields is either None or null for all the testing cases in this PR. I think we should have at least a test case where it's not None or null.

>  Yes, I agree I will add a test for the same.

**Review by nauh — APPROVED** [review]
LGTM. Please wait for approval from thinhha as well before merging the changes. Thanks.

File: clouddq/classes/dq_configs_cache.py
Comment (thinhha) [inline clouddq/classes/dq_configs_cache.py:546]: Can we please add an integration test for this behaviour?

File: clouddq/classes/dq_entity.py
Comment (thinhha) [inline clouddq/classes/dq_entity.py:393]: If the code got to this point, it is safe to say there are no corresponding Dataplex entity_id for the entity_uri, therefore all "dataplex_*" fields should be set to 'None' to ensure correctness.

File: clouddq/integration/bigquery/bigquery_client.py
Comment (thinhha) [inline clouddq/integration/bigquery/bigquery_client.py:212]: Can we please add an integration test for this behaviour?

File: clouddq/integration/bigquery/bigquery_client.py
Comment (thinhha) [inline clouddq/integration/bigquery/bigquery_client.py:245]: Is this effectively "STRING" for all Hive-partitioned tables?

File: clouddq/integration/bigquery/bigquery_client.py
Comment (thinhha) [reply clouddq/integration/bigquery/bigquery_client.py:240]: It is intuitive for a field to be either a STRING or an ARRAY. I would prefer if we keep it consistent as an ARRAY in the CloudDQ implementation.

Is there any reason why we cannot force '_PARTITIONTIME' into the ARRAY format? e.g.

    "partition_fields": [{"name": "_PARTITIONTIME",
                          "partitioning_type": "DAY", <- ensure you get the correct partition type of day, hourly, or monthly
                          "type": "TIMESTAMP"}]

Comment (thinhha) [reply clouddq/integration/bigquery/bigquery_client.py:240]: Could you please also add an integration test for _PARTITIONTIME time-partitioned table?

**Review by thinhha — COMMENTED** [review]
Adding a few more nits to be addressed in #157

File: clouddq/integration/bigquery/bigquery_client.py
Comment (thinhha) [reply clouddq/integration/bigquery/bigquery_client.py:245]: Does BigQuery tell us what the partition duration setting is? e.g. DAY, HOUR, MONTH

Comment (AmandeepSinghCS) [reply clouddq/integration/bigquery/bigquery_client.py:245]: Yes

File: clouddq/integration/bigquery/bigquery_client.py
Comment (AmandeepSinghCS) [reply clouddq/integration/bigquery/bigquery_client.py:212]: added the tests for this behaviour

File: clouddq/classes/dq_configs_cache.py
Comment (AmandeepSinghCS) [reply clouddq/classes/dq_configs_cache.py:546]: added the test for this behaviour

File: clouddq/classes/dq_entity.py
Comment (AmandeepSinghCS) [reply clouddq/classes/dq_entity.py:393]: Yes I agree. All this should be None by default if the code got to this point.

**Pull request merged**
